### ISSUE TITLE
Nitrokey Pro support

### ISSUE THREAD

**Issue by Germano0** [keepassxreboot/keepassxc#255]
[Nitrokey Pro](https://shop.nitrokey.com/shop/product/nitrokey-pro-3) is an Open Hardware device and [FOSS software](https://github.com/Nitrokey) that allows people to use such device as a secure token for various security needs, like two factor authentication, GPG/PGP, SSH keys, etc.
It would be great to have support for such device

**Comment by albrechtd** [comment]
I think this extension would be extremely useful and thus I strongly support this suggestion.

Maybe you know that there has been a somewhat similar [Feature Request for keypassx in 2010](https://forum.keepassx.org/viewtopic.php?t=2188).  The workaround using the gpg pipe doesn't work for me, though.

Instead of directly interfacing to a smartcard, an alternative could be a gpg-encrypted keyfile which is fed through gpgme which in turn uses the gpg key from any smartcard.

**Comment by droidmonkey** [comment]
I wish these devices used the PKCS#11 standard like smart cards.... I am not keen to include the libraries of all the various hardware tokens out in the world.

**Comment by Germano0** [comment]
@droidmonkey Nitrokey Pro should really be PKCS#11 compatible https://raymii.org/s/articles/Get_Started_With_The_Nitrokey_HSM.html#PKCS%2311,_%2315_and_OpenSC

**Comment by droidmonkey** [comment]
Hmmmm i wonder if we could just create a pkcs#11 module instead and then you could also use smart cards /pivs

**Comment by drdaeman** [comment]
I have a Gnuk-based FST-01, which should be quite similar to Nitrokey Start. I think while it works as OpenPGP Card v2, it doesn't really work as PKCS#11, at least not without something like Scute. I.e. `gpg --card-status` works out-of-box, but `pkcs11-tool --show-info` fails with `CKR_DEVICE_ERROR`. Maybe that's just me not understanding something, though.

How about an option of having an encrypted keyfile and talking to GnuPG (via GPGME or whatever), letting it handle the actual hardware, whatever it is? (Or maybe even a generic option of an external keyfile material provider using pipes?) Something similar in spirit to the linked KeepassX feature request, but with auth process explicitly started and controlled by KeepassXC.

**Comment by livelace** [comment]
@droidmonkey 

Hello. When do you plan to add support PKCS #11 ?

**Comment by droidmonkey** [comment]
Probably 2.3.0

**Comment by livelace** [comment]
@droidmonkey 

Thanks. I will wait with impatience.

**Comment by AzureRaptor** [comment]
I'm also interested in seeing any support, PKCS or otherwise, that would let us use NitroKeys.  I started to look at getting a YubiKey recently, but it seems like they are turning their backs on open source just lately, and their excuses are all very facile and shady.

**Comment by droidmonkey** [comment]
I noticed their GitHub is not very organized.

**Comment by rhetr** [comment]
https://github.com/Nitrokey/nitrokey-keepass-plugin they have a plugin repo for keepass with no documentation ✌

**Comment by v01d** [comment]
It would be also interesting for me to have support for Gnuk-based device. This would involve talking to gpg-agent for authentication. I understand this uses a standard protocol.

**Comment by phoerious** [comment]
We cannot use Nitrokeys directly, because they do not support HMAC-SHA1. We have some ideas for supporting PGP-encrypted keys, but it will take some time to implement then.

**Comment by szszszsz** [comment]
Hi! Just noticed this issue and I thought I could reply some questions regarding Nitrokey.

@droidmonkey Nitrokeys support PKCS#11 with OpenSC.

@drdaeman PKCS#11 with Nitrokey Start was used in [Raymii's tutorial](https://raymii.org/s/articles/Nitrokey_Start_Getting_started_guide.html) - maybe you would find this helpful. See guides from official Nitrokey's site as well.

@rhetr I do not know the details, but this is an attempt to use PKCS#11 in Keepass2. It perhaps worked 5 years ago, but I am not so sure about it right now. Good point to mention that in Readme.

@phoerious We support [HOTP](https://en.wikipedia.org/wiki/HMAC-based_One-time_Password_algorithm) at the moment. Maybe we could work something out with it, like using HOTP codes in encryption of the database (similarly to [Keepass2 solution](https://keepass.info/plugins.html#otpkeyprov))? 
In the future we could support HMAC-SHA1 directly, if there would be a request (no one registered it in [Pro's firmware project](https://github.com/Nitrokey/nitrokey-pro-firmware/issues) yet).

Love the project - all the best!

**Comment by livelace** [comment]
I'm still waiting PKCS11 :)

**Comment by phoerious** [comment]
We haven't had time for working on the PGP thing yet, there are too many other more urgent things that need to be done first, but it's on the agenda.
HOTP is basically just manual HMAC-SHA1. We could support it. I was also thinking to provide some sort of HMAC-SHA1 backup function where you enter the secret directly and let KeePassXC generate the challenge from it. That way you can still access the database if you lost your key, but still know its secret.

**Comment by jans23** [comment]
@phoerious Let me know if there is anything we at Nitrokey could do to support you.

**Comment by svenseeberg** [comment]
@jans23 I started working on this issue. However, libnitrokey has a steep learning curve (as has KeepassXC). Anyways, I do not understand how to connect to a Nitrokey while the Nitrokey app is running. Or put differently: NitrokeyManager->connect() only works while the Nitrokey app is not running. How do I solve this?

**Comment by szszszsz** [comment]
@svenseeberg Current protocol is not designed to be used for multiple applications thus opening in libnitrokey blocks further connections. Do you have a specific need to have both Nitrokey App and KeepassXC opened at the same time?

**Comment by svenseeberg** [comment]
@szszszsz Thanks for getting back to me that quick. No, there is no specific need. It's just about usability. Maybe I'll add a function that detects whether the app is running and notifies the user. And I guess, that the user could manually copy the one time pad if both programs are being used at the same time.

**Comment by szszszsz** [comment]
OK! Detecting solution as you speak should work (maybe supported with `libusb`). Perhaps adding such function directly to libnitrokey would be handy - if so, please register an issue [there](https://github.com/Nitrokey/libnitrokey/issues). Let me know if you would need any help with the library - I will happily support you. I have added myself to the fork watchers, so we could discuss issues there as well.

**Comment by svenseeberg** [comment]
Sorry for causing confusion. HOTP is no use because the shared secret would have to be stored in the kdbx file. So we're back waiting for PKSC11 support.

**Comment by gartfr** [comment]
Maybe have a look to this old keepass2 plugins which enables the possibility to store the master key which encrypts the DB to a pkcs11 token.
I use it since a long time now working quite well.
It relies on the driver DLL which do the interface between the hardware token and the software.
Personnaly, I use ePass2003 token  and opensc lib.

Main drawback, it is actually not working under linux... only Windows.

I'm no coder, I can't help anymore... sorry...

[CryptokiKeyProvider_0.1.3.zip](https://github.com/keepassxreboot/keepassxc/files/2416289/CryptokiKeyProvider_0.1.3.zip)

**Comment by maufl** [comment]
Hey, I just found this issue. What I would love is PKCS#11 support and support for encrypting it for multiple public keys. We use smart cards at work and it would be awesome if we could share password databases.

**Comment by droidmonkey** [comment]
This is good for *nix environments. For Windows you would want to interface directly with the certificate store so that third party PKCS#11 solutions (such as ActivClient) or even Windows itself can perform the interactions with the smart card. 

https://docs.microsoft.com/en-us/windows/desktop/secauthn/smart-card-authentication

**Comment by jans23** [comment]
To clarify: Windows' certificate store is different than PKCS#11. GnuPG supports only devices compatible with OpenPGP Card (e.g. Nitrokey Start, Pro, Storage) but not PKCS#1. You can choose to support PKCS#11 and/or Windows and/or GnuPG/OpenPGP Card.

**Comment by phoerious** [comment]
As mentioned earlier, I had some ideas about integrating OpenPGP with KeePassXC myself. Personally, however, I would prefer not to paste the keys into KeePassXC, but use the gpg-agent interface for encryption and decryption. That provides a lot more transparency and KeePassXC would be fully agnostic of the underlying hardware, keys, and algorithms and we do not have any extra hard dependencies in our code.

My idea was that if a database is open, you can "add" a PGP key by clicking a button in the database settings, which generates a random secret and lets the gpg-agent encrypt it. The encrypted secret is then stored in the KDBX public custom data block together with other keys by other people. For backup purposes, we can also display the plain secret to the user afterwards. Generally, the database settings allows you to add or remove any number of PGP-encrypted secrets. As an additional feature, one could generate an additional encrypted secret to be shared on a secure authenticated channel with a third party, so they can add it to their database so as to give you access without the need for your own private key.

**Comment by jans23** [comment]
HMAC-SHA1 was mentioned earlier. Could somebody describe what exactly is required in this respect? Is this a standard or custom solution?

Note, Nitrokey supports standard HOTP and TOTP.

**Comment by phoerious** [comment]
It's just standard HMAC-SHA1. Nothing special. I don't know about the actual protocol for talking to the USB device, though. That is handled by the YubiKey lib.

**Comment by jans23** [comment]
Perhaps HOTP would be a better alternative to HMAC-SHA1 because it is standard-compliant. However, PKCS#11 is still a more secure alternative.

**Comment by svenseeberg** [comment]
How would HOTP be used? AFAICT it requires a shared secret and this shared secret would have to be stored in the KDBX file. Therefore it provides no additional security. Am I getting something wrong?

**Comment by jans23** [comment]
Yes, HOTP uses a shared secret. I was assuming that HMAC-SHA1 works with
a shared secret too. Doesn't it?

**Comment by szszszsz** [comment]
Hi!
@svenseeberg If I got this right (and as long as I remember), in the original solution [1] the HOTP code was either part of the final secret used to encrypt the database, or a key for decryption of the semi-stage secret, which was further used with the final secret to open the database. The former requires rewriting the database file due to re-encryption, the latter - only re-encrypting the semi-password stored in the header I guess.
[1] is open source, so it should be easy to confirm that.

[1] https://keepass.info/plugins.html#otpkeyprov

**Comment by beelze** [comment]
Voting for pkcs#11 support. Something alike in Firefox, with ability to use custom (e.g. proprietary) pkcs#11 libraries

**Comment by rugk** [comment]
IMHO we should split this issue to "Nitrokey Pro support" and "PKCS#11 support" as it is really not clear whether PKCS#11 support would work for Nitrokey and solve the issue.

The thing is: Whatever implemented, it **should be compatible** with other KeePass-compatible client applications.
And actually, there is already something done and both solutions seem to use HOTP instead of PKCS#11:
* as already shown, [Nitrokey's own doc recommends a plugin for KeePass](https://www.nitrokey.com/documentation/applications#a:password-manager) called [otpkeyprov](https://keepass.info/plugins.html#otpkeyprov)
* also [Password Safe](https://gitlab.gnome.org/World/PasswordSafe) had the [request to support NitroKey](https://gitlab.gnome.org/World/PasswordSafe/issues/42) and [**already started** implementing support via HOTP](https://gitlab.gnome.org/World/PasswordSafe/merge_requests/53).

So as I see it, PKCS#11 support is not the thing other clients want to use for Nitrokey support. They all seem to use HOTP. So why exactly do you think it is not possible/secure here?

**Comment by droidmonkey** [comment]
Just because people are using HOTP doesn't make it the right or desirable answer.

**Comment by rugk** [comment]
Obviously yes, but we should tell these people this, then. The question is then, why is not it the right answer?

**Comment by droidmonkey** [comment]
HOTP requires a rider file in addition to your database file OR the shared "semi-secret" stored in the unencrypted portion of the database file. Read the posts from svenseeberg and szszszsz above. This has also been discussed elsewhere on the issue board. 

Read more here: https://en.wikipedia.org/wiki/HMAC-based_One-time_Password_algorithm

HOTP works great for client-server type relationships. It does NOT work great for encrypted data at rest (ie, KDBX)

**Comment by rugk** [comment]
"rider file"?

**Comment by droidmonkey** [comment]
An extra file in addition to your database file.

**Comment by rugk** [comment]
So that's the only issue? I mean, as long as it is cryptographically secure I see no problem with that, it's okay if it needs some extra data. In [here](https://github.com/keepassxreboot/keepassxc/issues/255#issuecomment-450149700) it is mentioned it may also need to re-encrypt/rewrite some data, but especially if it is just the header, I think there is no problem if it is an one-time step. (and when people setup their Nitrokey I guess they may also wait some seconds)

**Comment by gartfr** [comment]
I agree with the previous post, OTP is only adding the ability to avoid brute force attack and a factor of authentication (something you own).
PKCS#11 gives the 2 previous, but also gives the ability to store a complex encryption key on a secured device.
To me, PKCS11 support in conjunction with a USB Token or SmartCard is the ultimate security for Keepass.
Keepass security relies mainly on the hardeness of the database key which is used for encryption. PKCS11 adds the ability to store a very complex key on a device which will take care of the security of this precious encryption key : hardware encrypted storage, unlocked by pincode or more secured biometric function.

**Comment by phoerious** [comment]
The way HOTP is implemented in KeePass is that you store a secret in a sidecar file and encrypt it a couple of times with a series of one-time passwords. The scheme works, but is a hack at best. Six- or eight-digit numbers are quite easy to brute force, so the protection you gain depends a lot on the number of OTPs you generate. Also since those OTPs are not actually independent of each other, the actual entropy is much lower than that of a random string of the same length.

**Comment by cmosek** [comment]
Voting for HSM/SmartCard support via PKCS#11. 

This is the last thing that is stopping me from Keepass migration. Currently using [CertKeyProvicer](https://github.com/markbott/CertKeyProvider) plugin which is accessing Microsoft CryptoAPI (CAPI).

**Comment by droidmonkey** [comment]
In the meantime, if you really want to switch, I recommend just buying a Yubikey and using that as your token. It has the same level of protection as a smartcard.

**Comment by onlykey** [comment]
I found this issue while doing some research for another issue where we are implementing open source challenge and response https://github.com/keepassxreboot/keepassxc/issues/2064 and thought I would give my 2 cents.

HOTP - This value changes every time so its not clear to me how this will be used to encrypt a database. Sorry I didn't read all the comments above maybe there is a way. You could definitely store HOTP seeds in the encrypted databases for accounts but that is something different. 

Challenge-response (HMACSHA1) - This value does not change every time it is dependent on input. So you can send a challenge, get a response, use the response to encrypt database. You can also store the challenge in plain-text with the encrypted database as its not sensitive. To decrypt you send the challenge, get the same fixed response to decrypt. This is pretty good as the threat model here is an attacker would have to have root access on your computer and in real time capture the USB data containing the fixed response. One thing here is that the HMACSHA1 input should change often to mitigate threat. 

OpenPGP - Here you would encrypt the entire database with a private key. Usually what is done is there is an RSA/ECC key stored on a USB device (smartcard or token), that key is used to encrypt a symmetric key (32 bytes AES-256), that symmetric key is used the encrypt the database. This is basically how GPG/PGP works. This is a little better than challenge-response as the threat model here is an attacker would have to have root access and memory access to where the decrypted database resides. Each time the database is encrypted a new symmetric encryption key is used.

One additional threat model to consider with smartcard/token is that you are entering your PIN to unlock it on your keyboard. In the event that there is a keylogger on your computer and you forget to  remove the device an attacker can unlock your smartcard/token and use it as they will. This is one area where challenge-response has the advantage as you can require physical button press on the device. This threat model is one reason why [OnlyKey](https://onlykey.io) has the PIN entry on the device itself.

**Comment by phoerious** [comment]
HOTP does not "change every time". It simply depends on a secret and a counter. So given a fixed secret and a fixed counter, the generated OTP is also fixed.

**Comment by onlykey** [comment]
@phoerious OK I see what you are trying to do. If you are able to specify to the token the counter to use it would be fixed, however I thought for security reasons tokens always keep this counter internally. Either way, even if this works a fixed 6-8digit OTP is trivial to brute force so I am not sure what benefit this would offer.

**Comment by jans23** [comment]
I think bruteforce may be the reason why other password managers require three subsequent HOTPs in order to unlock the password safe.

**Comment by phoerious** [comment]
The HOTP KeePass plugin calculates a few OTPs in advance. The HOTP client then needs to know the counter value. It's very hacky.

**Comment by szszszsz** [comment]
Yes, the plugin requires three HOTP codes AFAIR, which gives 18+ additional digits (starts to be a problem, if your HMAC iterations takes 1 second, as recommended), but that's a detail.

One way or another, I think we should skip discussing the HOTP idea to not waste time, and focus on implementing another universal, but more secure solution and based on smart card, like PKCS#11. Limiting only to OpenPGP will cross out other smart cards, like HSM, hence the idea of using OpenSC. However, if it would be implemented as generic as possible, we could add OpenSC support later, and now use the GnuPG, as you prefer to.

@svenseeberg How is progress going? Could I help with testing?

**Comment by droidmonkey** [comment]
This should (and will) be compatible with US Government Common Access Cards as well. It would certainly make it easy and efficient to secure all your passwords in a database secured with your already-in-use CAC for logging into the machine.

**Comment by svenseeberg** [comment]
Not too much progress yet, sorry. I created a basic outline of how I imagine it can work. But my time is somewhat limited right now. But please feel welcome to make suggestions in my WiP here: https://github.com/svenseeberg/keepassxc/pull/1/files I'll gladly take up any useful suggestions.

**Comment by svenseeberg** [comment]
@phoerious (picking up our discussion in #2789) AFAICT gpg-agent only allows using known Smart Cards (and it still confuses me how to exactly use it). I'd prefer to directly use GnuPG (i.e. via GPGME) because that allows, for example, encrypting the secret with any available public key. But I'm currently also considering implementing both with different compile flags.

**Comment by alex-nitrokey** [comment]
Unfortunately, interacting with smart cards is not implemented in GPGME yet.

**Comment by Bjoe** [comment]
I've just studied how to add Nitrokey support into KeepassXC but it is good to read that are somebody is working on this :-)

It is good to hear that KeepassXC get now GnuPG support, thanks @svenseeberg for your contribution. And maybe I find some time to test your branch ...

The problem that I have with GnuPG is the scdaemon! Every time when I use my Nitrokey with the Nitrokey APP, scdaemon is going crazy and stuck, because they didn't close the connection.
I have to "pkill -9 gpg-agent" every time and wait 2 min. or I must unplug my Nitrokey and plug in again ... it's really annoying!

I agree completely with @szszszsz comment https://github.com/keepassxreboot/keepassxc/issues/255#issuecomment-464855428 . So I do not need to use the scdaemon :-)

But I have to adapt the statement a bit, KeepassXC should have PKCS#11 support not OpenSC. OpenSC provides only a PKCS#11 modul/library.
For more details see discussion in [smartcard support for NeoPG](https://github.com/das-labor/neopg/issues/54)

But before KeepassXC implement a PKCS#11 interface, there are already exists some libs to implment PKCS#11 support.

I know two:

* [pkcs11-helper](https://github.com/OpenSC/pkcs11-helper)
  It is a small library written in C. I didn't try this library, I'm a C++ programmer :-) and the API implementation of PKCS#11 is a very low level.
  
* [botan](https://botan.randombit.net/manual/pkcs11.html)
  I really recommend this library. It is written in C++ and the API implementation of PKCS#11 is very good and easy to understand. 

I already create a [POC version](https://github.com/Bjoe/pkcs11-botan-poc) for NeoPG with botan and PKCS#11 and you can play around with that to get some idea how this works. I currently plan to implemnt this in NeoPG.

If you need some help to implement PKCS#11 support, let me know.

And some additional note:
Yes, botan is a [Crypto library](https://botan.randombit.net/) like gcrypt. Maybe you will switch to botan in the future because I find the API more intuitive to use then the gcrypt API. But this is a discussion in a new "issue"/wishlist...

**Comment by droidmonkey** [comment]
Wow botan looks amazing. If they had argon2 kdf we could replace libsodium and argon2 with botan entirely.

**Comment by svenseeberg** [comment]
Thanks, I'll look into botan!

**Comment by randombit** [comment]
@droidmonkey Someone just pointed me at this thread. I'm the upstream maintainer of Botan. I just wanted to let you know that Argon2 was recently added and will be included starting in the next release (early July). We'd be very happy to add KeePassXC as a user so if there are any changes or improvements we could make that would make a switch easier or more compelling please let me know either via GH issue on the Botan repo or an email to me (address in profile).

**Comment by droidmonkey** [comment]
The KDBX standard calls for Argon2d, which libsodium does not support. That would be the most important inclusion. We support AES-256, ChaCha20, and Blowfish stream ciphers.

**Comment by randombit** [comment]
Interesting, didn't know that about libsodium. Botan has full 2d/2i/2id support. We also support AES, ChaCha and Twofish (among others) with all common modes (CBC, CTR, GCM, XTS, etc).

**Comment by droidmonkey** [comment]
And pkcs#11

**Comment by phoerious** [comment]
Is Botan peer-reviewed? I am very hesitant to build upon crypto libraries that are not widely used and analysed.

**Comment by randombit** [comment]
@phoerious It was audited by a team in 2015 who found some issues (which were all resolved) https://botan.randombit.net/releases/audit_1.11.18.pdf

Various bug reports have also been provided by different researchers and crypto engineers. I cannot know how carefully any one of them checked the whole library though. You can see all reported security issues https://botan.randombit.net/security.html

In 2017 it was reviewed and approved by the German BSI (Federal Office for Information Security) for government use https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Krypto/Projektzusammenfassung_Botan.html and is used by open source projects such as strongSwan, ISC KEA, and Shadowsocks-qt5, and companies including Rockwell Automation, Panasonic, Mazda, IBM, Bosch, PSPDFKit, and Rohde & Schwarz among others https://github.com/randombit/botan/wiki/Users

Also FWIW I build and review cryptographic systems for a living. I have contributed changes to other libs including OpenSSL and mbedtls, and currently maintain the crypto code used in a FIPS validated HSM.

If there is any other information I can provide please do ask.

**Comment by droidmonkey** [comment]
Let's take the Botan discussion over to the new issue ^^

**Comment by phoerious** [comment]
Okay, that sounds not too bad. I'm just worried that libraries like this could go the way of libssh etc. where fundamental vulnerabilities stay undetected for years.

**Comment by creidhne88** [comment]
SO, going back to the PKCS#11 support. Any updates on it? =)

**Comment by svenseeberg** [comment]
Sorry, my branch is still only work in progress. And I did not find time to work on it for several months. And I cannot guarantee that I will find spare time soon.

**Comment by svenseeberg** [comment]
Is anyone else on 36c3? I'd be willing to spend one or 2 days there on a sprint for this feature. Some company would be very motivating.

**Comment by muellermartin** [comment]
I will be there and would be happy to give feedback (but I'm not fit in C++ and not familar with the code base of KeePassXC, so my help will be somewhat limited).

**Comment by 4jNsY6fCVqZv** [comment]
@svenseeberg Thanks for your idea!
I was about to write to you and ask if it's possible to bring together a small group of people who can help you realize this very popular feature ;-)

**Comment by jans23** [comment]
I try to get some tickets and - if successful - will be at 36C3.

**Comment by alex-nitrokey** [comment]
I may join, too :)

**Comment by szszszsz** [comment]
@svenseeberg Sounds good! I won't be there this year, but I can support remotely if that would help.

**Comment by NikkyAI** [comment]
i will be there to cheer :P

**Comment by 4jNsY6fCVqZv** [comment]
@svenseeberg So where and when should the meeting at 36C3 take place? It would be nice if people could also join/support as remotes.

**Comment by svenseeberg** [comment]
As soon as the session form is available again, we could announce a (kickoff-) workshop for this on https://events.ccc.de/congress/2019/wiki/index.php/Form:Session

For communication we could simply use the official KeePassXC IRC dev channel: chat.freenode.net / #keepassxc-dev

**Comment by 4jNsY6fCVqZv** [comment]
Marvelous, can you do that?

**Comment by svenseeberg** [comment]
I'll try to remember :+1:

**Comment by 4jNsY6fCVqZv** [comment]
@svenseeberg session form seems to be back again!

**Comment by 4jNsY6fCVqZv** [comment]
Hi, quick question: Will there be a session this week to solve this issue?

**Comment by svenseeberg** [comment]
I'm in. However, I cannot send the self organized session form. Some fields are required but do not accept any input. Maybe I just don't understand how to fill them. Anyways: I'll be at https://36c3.c3nav.de/l/netzbegruenung/. Feel free to join. I'll also be on IRC.

**Comment by 4jNsY6fCVqZv** [comment]
Experience shows that at least a fixed location & starting time helps. Any suggestions?
Maybe another person can try to fill in the form?

**Comment by svenseeberg** [comment]
Lets meet 2019-12-27 1800 CET at the above mentioned assembly. If required, we can meet a second time on 2019-12-28.

**Comment by jans23** [comment]
How is the progress? Should we make another session today or tomorrow?

**Comment by svenseeberg** [comment]
@jans23 some little progress. Meeting is always good. 1500 today?

**Comment by jans23** [comment]
OK. I'm at Critical Decentralization Assembly, table FOSSAsia. Meet there or somewhere else?
-- 
Sent from a mobile device.

**Comment by svenseeberg** [comment]
@jans23 I'll come to your assembly.

**Comment by svenseeberg** [comment]
I met with the Nitrokey team during 36c3 and we discussed how to implement this feature. Summed up OpenSC/pkcs11 is probably the best choice, for example by using the library https://github.com/OpenSC/libp11. However, the library just takes care of decryption (AFAICT). That means we need another dependency for encryption with RSA. We can use OpenSSL for that. Is KeePassXC already using another library that has the ability to do the encryption? Then there is the question if we should use PKCS11 as an alternative or addition to the normal password. Now I think that using it as an alternative is the best way. That means users can use the password as backup.

**Comment by droidmonkey** [comment]
The way KeePassXC and KDBX is structured, at the moment, is with a combined key data (master key) that is transformed to produce the encryption key. It would be best to treat PKCS11 similar to Yubikey Challenge-Response where you provide a random token to the PKCS11 device which returns a transformed/signed version that is then added to the master key material.

**Comment by droidmonkey** [comment]
Added $200 bounty, get to work guys!

**Comment by svenseeberg** [comment]
I'll add another $150 if someone finishes this before I can ;-)

**Comment by 4jNsY6fCVqZv** [comment]
@droidmonkey @svenseeberg Thanks for your bounties! Are there already people who can imagine to work on a solution now?

**Comment by banym** [comment]
I am interessed in FIDO U2F / FIDO2 support for the Nitrokey FIDO2 and would add $150 to the bounty if it be implemented to make it possible to use that key as second factor.

**Comment by droidmonkey** [comment]
There is another issue tracking FIDO2

**Comment by 4jNsY6fCVqZv** [comment]
@banym Which issue would you like to support with your Bounty? Not sure I got it right.

> -> Sorry added the bounty to the other thread since this is the wrong thread as pointed out.

https://github.com/keepassxreboot/keepassxc/issues/255 or https://github.com/keepassxreboot/keepassxc/issues/3560?

**Comment by 4jNsY6fCVqZv** [comment]
@banym Which issue would you like to support with your Bounty? https://github.com/keepassxreboot/keepassxc/issues/255 or https://github.com/keepassxreboot/keepassxc/issues/3560? Not sure I got things right.

@droidmonkey This issue is about Nitrokey Pro and PKCS#11 support. The other issue is about FIDO2 support.
What about support for the other Nitrokey models? Start, Storage, FIDO U2F...
Is support for these models already included in one of the two issues?

**Comment by droidmonkey** [comment]
This issue requires PKCS#11 integration, see my comment here: https://github.com/keepassxreboot/keepassxc/issues/255#issuecomment-569516777

**Comment by banym** [comment]
@4jNsY6fCVqZv since this one is for PKCS#11 I would support the other #3560 for fido2 integration. I updated my initial post here but didn't want to delete it in this one for transparency reasons.

**Comment by 4jNsY6fCVqZv** [comment]
@droidmonkey Then the last $150 should also be announced for https://github.com/keepassxreboot/keepassxc/issues/3560 and removed here again, right?

**Comment by droidmonkey** [comment]
I cannot change a bounty, that is done through the Bountysource service. I don't think you can move a bounty once placed.

**Comment by 4jNsY6fCVqZv** [comment]
Does this mean that there is no way to correctly allocate the 150 euros now? Or just not by you?

**Comment by phoerious** [comment]
Please contact Bounty Source.

**Comment by 4jNsY6fCVqZv** [comment]
It's not my bounty.
Maybe @banym would like to?

**Comment by banym** [comment]
I talked to our team at work even we are primarily interested in FIDO2, we would like to see PKCS#11 for other use cases, too. So I decided to keep the $150 here, too. 

No need to change the the bounty here.

**Comment by 4jNsY6fCVqZv** [comment]
thanks for getting back!

**Comment by 4jNsY6fCVqZv** [comment]
thanks for getting back and supporting this issue!

**Comment by psiegl** [comment]
OpenSSH does now support FIDO2: https://www.phoronix.com/scan.php?page=news_item&px=OpenSSH-8.2-Released would this be an option to start with?

**Comment by prairiehare** [comment]
I add 100 EUR if you add support for Nitrokey Pro and PKCS#11

**Comment by rugk** [comment]
Regarding [the current campaign for SoloKey v2](https://www.kickstarter.com/projects/conorpatrick/solo-v2-safety-net-against-phishing) there has been some comment by the manufacturers on how that should be implemented:
> We still think the way to do this is to use the `hmac-secret` extension that is part of the actual standard (it is in essence HMAC-SHA256 challenge-response). In that sense, the answer is, yes, at launch (and Solo V1 has it too, just like every fully certified FIDO2 dongle).
> If the question is whether we will implement Yubico's proprietary app (in essence, HMAC-SHA1 challenge-response), the answer is no, we do not plan to do so. This sentiment is shared by other vendors such as Trezor - if OSS such as KeePassXC and LUKS moves on to an actual standard instead of one company's private sauce, everyone wins.

**Comment by rugk** [comment]
Ah sorry, just saw this is apparently https://github.com/keepassxreboot/keepassxc/issues/3560

**Comment by droidmonkey** [comment]
BTW, with my conversion of keepassxc to using Botan we are MUCH closer to integrating PKCS#11 in a standard, cross-platform manner. I anticipate this being implemented for 2.7.0 along with the Ledger key support.

**Comment by shufps** [comment]
Are there any news about PKCS#11 support? :innocent:

**Comment by jornfranke** [comment]
I would be also curios. One simple way could be also to add an option to add one or more gpg keys (I expect that one user has multiple openpgp smartcards in case one is lost) using the command line version of gpg. Then you have already multiple smart integrations out of the box. The users would just need to install gpg, but this can be optionally. I would think that most of the people with Nitrokey Pro or other openpgp smartcards have gpg already installed for sure.

**Comment by smesguich-orange** [comment]
Hello !
Very intrested of this feature !

**Comment by florianutz** [comment]
@droidmonkey 
> BTW, with my conversion of keepassxc to using Botan we are MUCH closer to integrating PKCS#11 in a standard, cross-platform manner. I anticipate this being implemented for 2.7.0 along with the Ledger key support.

Do you have any idea when it will be implemented? We are now at 2.7.1

**Comment by yan12125** [comment]
(From https://github.com/keepassxreboot/keepassxc/issues/255#issuecomment-569516777)

> It would be best to treat PKCS11 similar to Yubikey Challenge-Response where you provide a random token to the PKCS11 device which returns a transformed/signed version that is then added to the master key material.

I managed to sign data using Botan and PKCS#11 after reading some docs/examples. I don't have time to push it further recently, so I paste my current progress below. Hope it helps :)

<details>
<summary>
Codes that signs data with an RSA private key.
</summary>

```C++
#include <cstdint>
#include <string>
#include <iostream>
#include <iomanip>
#include <botan/auto_rng.h>
#include <botan/pubkey.h>
#include <botan/p11_rsa.h>
#include <botan/p11_types.h>
#include <botan/p11_object.h>

// echo -n yay | openssl dgst -engine pkcs11 -keyform ENGINE -sign 'pkcs11:token=OpenPGP%20card%20%28User%20PIN%29;id=%03;object=Authentication%20key;type=private' | od -x --endian=big
int main() {
    Botan::PKCS11::Module pkcs11_module("/usr/lib/pkcs11/opensc-pkcs11.so");
    std::vector<Botan::PKCS11::SlotId> slots = Botan::PKCS11::Slot::get_available_slots(pkcs11_module, true );
    // use first slot
    Botan::PKCS11::Slot slot( pkcs11_module, slots.at( 0 ) );
    Botan::PKCS11::SlotInfo slot_info = slot.get_slot_info();
    std::cout << std::to_string( slot_info.firmwareVersion.major ) << "."
              << std::to_string( slot_info.firmwareVersion.minor ) << std::endl;
    Botan::PKCS11::Session read_only_session( slot, true );
    Botan::PKCS11::secure_string pin = { '1', '2', '3', '4', '5', '6' };
    read_only_session.login( Botan::PKCS11::UserType::User, pin );
    const auto& objs = Botan::PKCS11::Object::search<Botan::PKCS11::PKCS11_RSA_PrivateKey>(read_only_session, std::string("Authentication key"));
    std::cout << objs.size() << std::endl;
    if (!objs.size()) {
        std::cout << "No objects found!" << std::endl;
    } else {
        const Botan::PKCS11::PKCS11_RSA_PrivateKey& obj = objs.at(0);
        for (uint8_t c : obj.get_attribute_value(Botan::PKCS11::AttributeType::Id)) {
            std::cout << std::hex << std::setfill('0') << std::setw(2) << static_cast<uint16_t>(c) << ":";
        }
        std::cout << std::endl;

        Botan::Null_RNG rng;
        // EMSA4 for RSA uses PSS (probabilistic signature scheme)
        Botan::PK_Signer signer( obj, rng, "EMSA3(SHA-256)", Botan::IEEE_1363 );
        Botan::PKCS11::secure_string plaintext = { 'y', 'a', 'y' };
        const auto& signature = signer.sign_message( plaintext, rng );
        for (const unsigned char c : signature) {
            std::cout << std::hex << std::setfill('0') << std::setw(2) << static_cast<uint16_t>(c) << ":";
        }
        std::cout << std::endl;
    }

    read_only_session.logoff();
}
```
</details>

Tested with Nitrokey Start. Here I use an RSA key instead of an Ed25519 key as Botan API does not support it yet. [Botan docs](https://botan.randombit.net/handbook/api_ref/pkcs11.html) mentions

> The Botan PKCS#11 interface is implemented against version v2.40 of the standard.

while Ed25519 support is added in [PKCS#11 v3.0](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html)

**Comment by droidmonkey** [comment]
I thought of a major problem with pkcs#11 and the nature of PKI which is NOT meant to be static. When your cert expires and you get a new one or a new card you will be locked out of your database. So unless we are strictly doing certificate validation... PKCS#11 is not a good choice for an offline database.

**Comment by yan12125** [comment]
I think I only need the private key to sign data for using as a keepassxc master key material, and a certificate is not needed? I don't use a certificate in the example above. I can keep the current master key as long as I keep the PKCS#11-capable hardware key, just like how KeePassXC works for Yubikey-like keys.

**Comment by My1** [comment]
Can't pkcs11 also be used to en/decrypt stuff? Because with signing there might be a chance there's some randomness added or whatever for security which would 100% break this idea. Also not sure if there can be certs or just keys which don't even expire. And in fact expiry doesn't really matter even if keypass checks it as you need to keep in mind that the database is already in hands of the attacker so they could just decrypt it manually after getting the signature/decrypted value.

Any kinds of 'smart play' would basically amount to snake oil.

**Comment by yan12125** [comment]
The [PKCS#11 v3.0](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html) standard does mention encryption/decryption. I believe Botan also supports that but I don't have time to verify it recently.

**Comment by droidmonkey** [comment]
Encrypt/decrypt and signing would produce the same failure if the underlying private key changes. You also don't want to encrypt anything of large size with the on-board chip of a pkcs#11 device.

**Comment by My1** [comment]
yes if the key changes there'll be a problem. simple solution, dont change that key. (except in a scenario where you have the DB open for re-keying obviously)
I dont know all smartcards obviously but those that I know usually have multiple slots so set one up specifically for keepass to just not change.

the problem is any scenario where you wanna have the ability to change your key with basically no notice requires a method to disallow the data to be processed/transmitted/whatever in a robust manner, purely on a verify, basically a data valve.

This however only works where you actually can stop the data flow in that manner. for example when you sign into your nextcloud, your nextcloud could deny sending you the data.

where a verify-based approach does NOT work is when you already have the data to be processed and the process is known, say a password database, a master password and an open source password manager.

because you could either fork the program and patch out the verify, do the decryption yourself, or if you are really crazy, just knock out the verify in RAM in case the program isnt open source.

It is the same reason why you [should not](https://blog.my1.dev/steganos-privacy-suite-19-is-a-joke) use [TOTP for data/password vaults](https://www.steganos.com/en/products/steganos-privacy-suite). (yes this is literally 5 years old and they are still doing it XD)

**Comment by livelace** [comment]
It's worthy to check out systemd pkcs11 support implementation - https://www.freedesktop.org/software/systemd/man/systemd-cryptenroll.html

I use Rutoken HSM for unlocking LUKS, as soon as my certificate will finally expired (10 years, btw) - I have an option to unlock LUKS with regular passphrase (backup mechanism).

The main point of using HSM - don't let passphrase to be leaked, keep secrets in a safe place.

**Comment by droidmonkey** [comment]
> yes if the key changes there'll be a problem. simple solution, dont change that key.

This only applies in cases where you have control over your key. If you use an employers smart card or a government card there is no guarantee your key won't change if you lose your card or have to get it renewed. Also if you are using certificates your key will expire at some point and will be changed.

**Comment by My1** [comment]
> This only applies in cases where you have control over your key.

either that or the person who does have control is aware and/or you can work with them about that.

> If you use an employers smart card or a government card there is no guarantee your key won't change if you lose your card or have to get it renewed.

if it's not even your card then maybe it's not the best idea to use it. Also, if your employer wants you to use the card for keypass there are ways to get around:

1) have the employer have a backup of that key (especially useful if needed to provision that to multiple ppl anyway for example in case of a team safe, in fact that's also a great way if you wanna ensure the people who can access the safe dont have access to the key on the card itself but are only able to use it, take the smartcard away and the access is revoked by at worst rotating the DB master key while still allowing the smartcards with the key to unlock seamlessly)
2) have a backup of the key yourself and get it onto that card if the employer allows that
3) have a backup procedure to get into the safe for rekeying

> Also if you are using certificates your key will expire at some point and will be changed.

That's technically not true. the cert expires and you can make a new cert onto the same key if you still trust it.

also as said smartcards can have multiple slots so you can use one like a proper cert with expiry rekeying and all and another slot just for keepass.

**Comment by ffries** [comment]
+1

I am using OpenSC smartcards and any OpenSC smartcard should do the job.
OpenSC is supported on all systems. It is today's world standard.

OpenSC smartcards allow to store an AES/GPG key and retrieve it with PIN protection.
But the best would be to keep the AES key secure and derive the key using the smartcard.

https://www.smartcard-hsm.com/features.html#smartcard
> EC Key Derive
> In particular for crypto currencies and smart contracts, the SmartCard-HSM supports derived EC keys to implement key management schemes similar to BIP32. Unlike with BIP32, key material remains inside the secure element, while the public key can be calculated outside from the master public key.

The Smartcard-HSM card works pretty well and is probably the most advanced OpenSC smartcard available, is both contact and contactless (and the cheapest):
https://www.smartcard-hsm.com/

There are also several open-source implementations of OpenSC smartcards.
Example: French National Security Agency OpenGPG smartcard  (compatible with OpenSC): 
https://github.com/ANSSI-FR/SmartPGP

The advantage of smartcards over USB keys is that they can be used with PINPAD readers.
A good CCID pinpad reader with secure messaging (don't buy without secure messaging) costs 30€ and is unbreakable.
secure messaging makes sure that PIN code is transmitted securely beetween pinpad and reader.

Example :  SCM SPR332 : v1 around 20€ used and v2 around 50€ new.
https://www.floss-shop.de/en/security-privacy/smartcard-reader/26/scm-spr332-v2-pinpad-reader

Any CCID pinpad is supported out of the box without need for drivers on all systems.
See CCID website for reference: https://ccid.apdu.fr/

Unfortunately Yubikey is a nice product name but offers no pinpad so no real security.
Same as for Nitokey, it is not real security.

**Comment by ffries** [comment]
By the way, I just discovered that the Nitrokey is based on the smartcard-hsm.
The Key / smartcard offers full support for PKCS#11 on Linux, Mac and Windows using OpenSC.

https://www.smartcard-hsm.com

OpenSC offers rock-solid APIs and PKCS#11 libraries.

https://github.com/OpenSC/OpenSC

So it should be the way to go ...

hmac-sha1 is deprecated. The PKCS#11 supports several ways to derive a key and/or calculate a SHA-256 challenge-response.

This thread should be renamed to: Add PKCS#11 support for smartcards encryption/decryption/signature

**Comment by frankmorgner** [comment]
Actually, you want to use a platform dependant method for using a token to minimize friction for the user.

For Windows, you should start here:
https://keepass.info/plugins.html#certkeyprov 
Use MS Crypto API to decrypt the master password. This will work for any token provider, not just OpenSC and it allows full compatibility with existing databases.

For macOS, you should use CryptoTokenKit and only on Unix you should use PKCS#11

**Comment by igor-krawczuk** [comment]
Hi, where can I add cash for this bounty? Is it simply pledging or is there a blessed platform like bountysource, issuehunt etc. that is being used?

**Comment by droidmonkey** [comment]
We don't endorse any bounty system anymore

**Comment by florianutz** [comment]
I dropped my money on bountysource two years ago but it didn't help this feature move forward.

**Comment by ncorder** [comment]
[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=41676090)](https://app.bountysource.com/issues/41676090-add-nitrokey-pro-and-pkcs-11-support)

________________________________

not entirely certain what the thought process on bounty source was here but wanted to add the widget

**Comment by smesguich-orange** [comment]
Hello,

I see that feature doesn't moove forward, I didn't get all why it's still stuck ? 
Also, bounty system is done ? So Bad ...

**Comment by rugk** [comment]
Support for NitroKey 3 (I assume it is the same product as the "NitroKey Pro 3") has been added in this PR apparently (just not linked):
https://github.com/keepassxreboot/keepassxc/pull/9631

So I guess it should be included in the next release… :eyes: :tada:

**Comment by florianutz** [comment]
@rugk 
> Support for NitroKey 3 (I assume it is the same product as the "NitroKey Pro 3") has been added in this PR apparently (just not linked): #9631
> 
> So I guess it should be included in the next release… 👀 🎉

#9631 is about challenge-response but not PKCS#11 what this thread is about

**Comment by jinnko** [comment]
For what it's worth, [OpenSC can use the openpgp inteface to access keys stored on an openpgp smartcard](https://wiki.archlinux.org/title/GnuPG#Multi_applet_smart_cards).

With that opensc config in place it's possible to load a GPG key from a smartcard such as a Yubikey into the ssh-agent with `ssh-add -s /usr/lib/pkcs11/opensc-pkcs11.so ~/.ssh/id_rsa-yubikey.pub`.

Could KeepassXC interface with the ssh-agent in a similar way to manage the key in the ssh-agent?

**Comment by Spamtributor** [comment]
I may be interested in implementing PKCS#11 support however it was not on my list of planned contributions for this project is the bounty still available?

Regardless, glad I just discovered the Nitrokey!

**Comment by droidmonkey** [comment]
To be quite frank, after diving deep on the Fido2 integration, doing PKCS#11 will suffer from the same problem. Derived key encryption is what KDBX uses. Both Fido2 and PKCS#11 require use of wrapped keys due to how they are implemented. Doing this with the current KDBX spec creates a huge mess. I DO NOT recommend working on this at all.

Also, there are no bounties.

**Comment by sschuberth** [comment]
> Also, there are no bounties.

Maybe adjust the title of this issue then, @droidmonkey?

**Comment by cen1** [comment]
Apologies for perhaps asking a dumb question, I am not strong in cryptography, but would it be possible to implement an "improved" HOTP over PKCS#11 using HMAC?

1. Have a master password and a `token_challenge` random nonce, which is stored in plain metadata.
2. Additional plain metadata field with Yubikey public fingerprint (to identify the device key).
3. Send instruction over PKCS#11 for Yubikey to calculate `H = HMAC(hardware_token, token_challenge)`.
4. Optional Yubikey interaction
5. `Final key = KDF(master_password + H)`. We treat H the same way as a static private key file option.
6. On each db change/save, regenerate `token_challenge`, compute new `H` and re-encrypt.

This way we don't need any raw key material from the hardware device (which is impossible) but we use the HMAC as an input into KDF.

**Comment by My1** [comment]
I think this would have no relation to PKCS11, but rather is just an expansion of the yubi HMAC Protocol usage to regularly rotate the key.

but rather than public FP (which wouldnt be ideal as you can have multiple with the same secret to actually work) you could make use of HMAC as a checksum with a known challenge and result and if those match you know you are talking to the right one.

the point of adding PKCS11 (or FIDO2 hmac-secret, as discussed in another ticket) is to open up compatibility away from just yubikeys, as also the future of the HMAC protocol is kinda unknown as the Yubikey Bio's Multi-Proto edition actually axes a lot of stuff including HMAC as far as I have read) and having PKCS11 support would make it usable with dang near every PIV thing you can imagine, be it nitrokeys, ordinary PIV cards that also exist, or even other USB PIV devices which are existing and/or upcoming, and yubikeys are still compatible with that as PIV has existed on the Yubi since the Neo from 2012.

**Comment by CardContact** [comment]
We are happy to fund the development of a PKCS#11 integration to support the [Nitrokey-HSM / SmartCard-HSM](https://www.smartcard-hsm.com).

IMHO the [key domain concept](https://www.smartcard-hsm.com/2025/09/30/key-from-a-to-b.html) would be great to manage the mobility of passkeys.

**Comment by phoerious** [comment]
@CardContact Thanks. Unfortunately, funding isn't as much of an issue as time right now. If you can hire someone to develop a pull request, we're generally happy to merge it. However, such a pull request should be integrated with the ongoing work on FIDO2 (which should be merged first I think).

**Comment by CardContact** [comment]
Is there a pull request for the FIDO stuff and does that use PKCS#11 ?

It would certainly need someone that knows the keypassxc code already to create a pull request. We can contribute the PKCS#11 part, either from OpenSC or our own module.

**Comment by phoerious** [comment]
#10311

**Comment by CardContact** [comment]
If I understand the pull request right, then it adds another authentication method, this time based on a FIDO2 hmac.

That is not what I was looking at. My use of PKCS#11 would be to decrypt / encrypt the database using an AES key on a hardware device via PKCS#11. Much like what we are doing with [file encryption](https://www.pki-as-a-service.org/km/fileencryption/) in native mode.

That way you can manage the encryption key using key domains, just like you would manage all other hardware-based cryptographic keys. Having passkeys on the hardware token would a bonus.

**Comment by droidmonkey** [comment]
That isnt how pkcs#11 works. The key never leaves the device and you cannot use the device to encrypt/decrypt things (that would be unbearably slow). This works same as hmac-secret, you provide a randomized seed that gets signed or encrypted by the device and the returned result is used as key material.

**Comment by My1** [comment]
couldnt one just encrypt the DB Master Key using the smartcard's cert and decrypt that when opening?
you could just for example have an absolutely insane master password you could throw into a safe and use in a pinch but for main usage just encrypt a copy of `T` ( https://keepass.info/help/kb/kdbx.html#keys ), and throw it right into `Public custom data` ( https://keepass.info/help/kb/kdbx.html#header )

heck you could even throw multiple encrypted `T`s into that thing to decrypt with multiple cards.

**Comment by phoerious** [comment]
> Much like what we are doing with [file encryption](https://www.pki-as-a-service.org/km/fileencryption/) in native mode.

The FIDO2 decryption would be more like the "derived" mode in the guide you linked above. That is also how PKCS#11 should be implemented here.

> couldnt one just encrypt the DB Master Key using the smartcard's cert and decrypt that when opening?

That's pretty much how it works.

> heck you could even throw multiple encrypted Ts into that thing to decrypt with multiple cards.

Same, though we don't store the already-derived secret. The keys from the public header would still go into the key derivation to keep things simple and consistent.

**Comment by CardContact** [comment]
> That isnt how pkcs#11 works. The key never leaves the device and you cannot use the device to encrypt/decrypt things (that would be unbearably slow)

PKCS#11 is a generic API for HSMs. What common tokens are implementing is often just a fraction of the standard - most commonly signing / decrypting with RSA keys or signing / key agreement with EC keys. All of that is not PQC ready, but SHA and AES are.

Throughput is at 5KB/s for contact-based communication and 10KB/s via NFC. My well filled KDBX is 6KB, so it would take a little over a second to decrypt the database.

And using a hardware base encryption key counters the harvest-now-decrypt-later thread, because asymmetric crypto is only used with XKEK key domains for key management, which could be restricted to a safe environment.

With derived mode you are still PQC ready and have the advantage, that the derived secret is based on a proper secret key, that you can manage in a key domain, which provides migration of keys between devices. Contrary to common believe, HMAC keys in secure elements are not treated the same way as RSA, ECC or AES keys. The later have special handling in the security controller, like encryption with a PUF. HMAC keys are just data elements in memory.

**Comment by phoerious** [comment]
> Throughput is at 5KB/s for contact-based communication and 10KB/s via NFC. My well filled KDBX is 6KB, so it would take a little over a second to decrypt the database.

KDBX files are often multiple MB. As soon as you start adding favicons and attachments, the size grows considerably. I have one with about 3.5MB. That would be impossible to decrypt that way.

Besides the speed concern, this would require a full redesign of how we encrypt and decrypt files. The only sensible way hardware-based encryption can be used is to encrypt/decrypt a secret on the crypto device, which is fed into the normal encryption/decryption routines on the host. Whether the device part is done by public key cryptography or symmetric AES is unimportant.

**Comment by CardContact** [comment]
> KDBX files are often multiple MB. As soon as you start adding favicons and attachments, the size grows considerably. I have one with about 3.5MB. That would be impossible to decrypt that way.

Yes, that is an issue we have with encrypting larger files as well. It's the reason why we have derived mode in our file encryption.

> Whether the device part is done by public key cryptography or symmetric AES is unimportant.

From the perspective of PQC-readyness I would challenge that. Of course PKCS#11 allows both variants.

**Comment by phoerious** [comment]
> From the perspective of PQC-readyness I would challenge that. Of course PKCS#11 allows both variants.

But that is up to the owner of the device to decide. In technical terms for how the KDBX is decrypted, both are identical.

> Yes, that is an issue we have with encrypting larger files as well. It's the reason why we have derived mode in our file encryption.

The current YubiKey implementation is pretty much this already. The only limitation is that you can enrol only a single HMAC secret. With FIDO2/PKCS#11, the encrypted secret would be stored in the public data, so multiple devices with different keys can be used.

**Comment by CardContact** [comment]
> The current YubiKey implementation

I wouldn't call that vendor-neutral.

**Comment by phoerious** [comment]
Not at all. It's even feature-gated behind specific USB VIDs and PIDs, which is why we're adding FIDO2 and are open to adding PKCS#11 as well. Just not with full on-device crypto for the payload.
